_posts/2018-09-13-refactoring-template-pattern.markdown

```diff
@@ -0,0 +1,107 @@
+
+<!--more-->
+
+## The Template Method
+
+The Template Method is described as <strong>_"a behavioral design pattern that lets you define the skeleton of an algorithm and allow subclasses to redefine certain steps of the algorithm without changing its structure."_</strong>
+
+The goal is to separate code that changes from code that doesn't change, keeping the concerns isolated on specialized classes. Those subclasses will them implement all the specific step.
+
+Check the image bellow that is illustrating the Template method structure:
+
+<div style="text-align: center; font-size: 12px; margin-bottom: 20px;">
+  <img src="/blog/assets/images/design-patterns/template-method.png">
+  <center><em>credits: Refactoring Guru (https://refactoring.guru)</em></center>
+</div>
+
+## Show me The Code
+
+Lets say that we are still building the employees application from our [last article](https://www.ombulabs.com/blog/code-refactor/refactoring-with-design-patterns.html) and now we need to sent an email to the managers with a report containing the amount of worked hours for each employee.
+
+The implementation of the `Report` class is quiet simple:
+
+```ruby
+class Report
+  def generate_report!
+    get_employees_worked_time
+    format_report
+    send_to_stakeholders
+  end
+
+  def get_employees_worked_time
+    # Retrieve this info from the database
+  end
+
+  def format_report
+    # Generate the HTML with the Report design
+  end
+
+  def send_to_stakeholders
+    # Call send email service
+  end
+end
+```
+
+That code works perfectly fine. But what if now we also need to generate the same report in text format? The only part that will vary is exactly the format report step, so that is a scenario when apply the *Template Method* is a good choice.
+
+To apply this pattern we will transform the `Report` class into an abstract class that can be inherited from several concrete classes.
+
+Back to our code, the only change necessary is to let the implementation of the format_report method for the children class:
+
+```ruby
+class ReportTemplate
+  def generate_report!
+    get_employees_worked_time
+    format_report
+    send_to_stakeholders
+  end
+
+  def get_employees_worked_time
+    # Retrieve this info from the database
+  end
+
+  def format_report
+    raise NotImplementedError
+  end
+
+  def send_to_stakeholders
+    # Call send email routine
+  end
+end
+```
+
+And for each variation of an report we need to create a concrete subclass:
+
+```ruby
+class HTMLReport < ReportTemplate
+  def format_report
+    # implement the report in HTML format
+  end
+end
+```
+
+```ruby
+class TextReport < ReportTemplate
+  def format_report
+    # implement the report in Text format
+  end
+end
+```
+
+Looks good, right? If we ever need to create a new format, we just need to create a new concrete class, making this a perfect example of the Open/Closed Principle. And by designing our code this way it will be easier and safer to change anything in the future.
```

Comment:
Maybe you could add a link to the Open/Closed Principle? I'm not sure what that is.